neqo-transport/src/connection/tests/datagram.rs

```diff
@@ -0,0 +1,243 @@
+}
+
+#[test]
+fn limit_data_size() {
+    let (mut client, mut server) = create_and_connect_with_dgrams();
+
+    assert!(u64::try_from(DATA_BIGGER_THAN_MTU.len()).unwrap() > DATAGRAM_LEN_MTU);
+    assert_eq!(
+        client.send_dgram(DATA_BIGGER_THAN_MTU),
+        Err(Error::TooMuchData)
+    );
+    assert_eq!(
+        server.send_dgram(DATA_BIGGER_THAN_MTU),
+        Err(Error::TooMuchData)
+    );
+}
+
+#[test]
+fn datagram_acked() {
+    let (mut client, mut server) = create_and_connect_with_dgrams();
+
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert_eq!(client.send_dgram(DATA_SMALLER_THAN_MTU), Ok(false));
+    let out = client.process_output(now()).dgram();
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent + 1);
+
+    let dgram_received = server.stats().frame_rx.datagram;
+    server.process_input(out.unwrap(), now());
+    assert_eq!(server.stats().frame_rx.datagram, dgram_received + 1);
+    let now = now() + Duration::from_millis(1);
+    // Ack should be sent
+    let ack_sent = server.stats().frame_tx.ack;
+    let out = server.process_output(now).dgram();
+    assert_eq!(server.stats().frame_tx.ack, ack_sent + 1);
+
+    let datagram =
+        |e| matches!(e, ConnectionEvent::Datagram(data) if data == DATA_SMALLER_THAN_MTU);
+    assert!(server.events().any(datagram));
+
+    client.process_input(out.unwrap(), now);
+    let datagram_acked = |e| matches!(e, ConnectionEvent::DatagramAcked);
+    assert!(client.events().any(datagram_acked));
+}
+
+#[test]
+fn datagram_lost() {
+    let (mut client, _) = create_and_connect_with_dgrams();
+
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert_eq!(client.send_dgram(DATA_SMALLER_THAN_MTU), Ok(false));
+    let _out = client.process_output(now()).dgram(); // This packet will be lost.
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent + 1);
+
+    // Wait for PTO
+    let now = now() + AT_LEAST_PTO;
+    let dgram_sent = client.stats().frame_tx.datagram;
+    let out = client.process_output(now).dgram();
+    assert!(out.is_some()); //PING probing
+                            // Datagram is not sent again.
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent);
+
+    let datagram_lost = |e| matches!(e, ConnectionEvent::DatagramLost);
+    assert!(client.events().any(datagram_lost));
+}
+
+#[test]
+fn datagram_sent_once() {
+    let (mut client, _) = create_and_connect_with_dgrams();
+
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert_eq!(client.send_dgram(DATA_SMALLER_THAN_MTU), Ok(false));
+    let _out = client.process_output(now()).dgram();
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent + 1);
+
+    // Call process_output again should not send any new Datagram.
+    let dgram_sent = client.stats().frame_tx.datagram;
+    assert!(client.process_output(now()).dgram().is_none());
+    assert_eq!(client.stats().frame_tx.datagram, dgram_sent);
+}
+
+#[test]
+fn dgram_no_allowed() {
```

Comment:
This is where you tweak the remote transport parameters.

I would use the test_frame_writer function for this instead.  You can write the DATAGRAM frame yourself.  It's trivial to do.

Reply:
I was searching for it and I could not find it. I knew we have something :)